### PR TITLE
Add tests for check_receipt in eth-relay

### PR DESCRIPTION
**Pull request by clearloop**
#274, #313 

## Steps
+ [x] mock headers
+ [x] canonical hash
  + [x] should succeed
  + [x] should fail
+ [x] block safe

## ~~DO NOT MERGE THIS PR NOW~~

~~The git history has some problems and they will be fixed at 20:00 tonight.~~

### CODE REVIEW

File: frame/chainrelay/eth/relay/src/tests.rs [hunk 1]
```diff
@@ -173,3 +173,58 @@ fn relay_header() {
...
+		//
+		// let mut gb = grandpa.clone();
+		// gb.author = EthAddress::from(hex!("bb5c52d9Ab611FCa798dDae930c7fc8307EfdC11"));
+		// gb.hash.take();
+		//
+		// assert_ok!(EthRelay::relay_header(Origin::signed(0), gb));
+		// assert_ok!(
+		// 	EthRelay::check_receipt(Origin::signed(0), receipt),
+		// 	<Error<Test>>::HeaderNC
+		// );
```
Comment (clearloop) [inline frame/chainrelay/eth/relay/src/tests.rs]: This `difficulty` part is really difficult, don't know how to mock, once I changed the info of a block, it's `difficulty` will slide down as I can't imagine.

File: frame/chainrelay/eth/relay/src/tests.rs [hunk 2]
```diff
@@ -1,175 +1,159 @@
...
-
-use eth_primitives::{
-	receipt::{LogEntry, TransactionOutcome},
-	Bloom, EthAddress, H64, U128,
-};
-use frame_support::assert_ok;
+use eth_primitives::{receipt::TransactionOutcome, U128};
+use frame_support::{assert_err, assert_ok};
 use frame_system::RawOrigin;
-use hex_literal::hex;
```
Comment (HackFisher) [inline frame/chainrelay/eth/relay/src/tests.rs]:
```
error: cannot find macro `hex` in this scope
   --> frame/chainrelay/eth/relay/src/tests.rs:181:15
    |
181 |             hash: Some(hex!("034cd83d9150808de742592f57e302dc9eccc71af270639dc5f236e5bdd7d3e3").into()),
    |                        ^^^
    |
    = help: have you added the `#[macro_use]` on the module/import?
error: cannot find macro `hex` in this scope
   --> frame/chainrelay/eth/relay/src/tests.rs:179:28
    |
179 |                 rlp::encode(&H64::from(hex!("a98d18400422c64e"))),
```

File: frame/chainrelay/eth/relay/src/mock_headers/utils.rs [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+use json::object::Object;
+use rustc_hex::FromHex;
+
+use crate::*;
+
+pub use eth_primitives::receipt::LogEntry;
+pub use json::JsonValue;
+
+/// hex
+macro_rules! hex {
```
Comment (AurevoirXavier) [inline frame/chainrelay/eth/relay/src/mock_headers/utils.rs]: Just use `<[u8: $len]>::from_hex($array_like)`

File: frame/chainrelay/eth/relay/src/tests.rs (hunk 2 above)
Comment (clearloop) [reply frame/chainrelay/eth/relay/src/tests.rs]: This is caused by the incorrect merge of the latest commit.

File: frame/chainrelay/eth/relay/src/mock_headers/utils.rs (hunk 3 above)
Comment (clearloop) [reply frame/chainrelay/eth/relay/src/mock_headers/utils.rs]: I'll replace the handwrite macro to rustc_hex

File: frame/chainrelay/eth/relay/src/tests.rs (hunk 2 above)
Comment (HackFisher) [reply frame/chainrelay/eth/relay/src/tests.rs]: Oops, still have compile error.

Comment (clearloop) [reply frame/chainrelay/eth/relay/src/tests.rs]: It's strange, the error part shoud have been removed alreadly, I'll find out why travis running the old code tonight.

Comment (clearloop) [reply frame/chainrelay/eth/relay/src/tests.rs]: Travis passed the tests on my [fork][0], but not at this pr.

[0]: https://github.com/clearloop/darwinia

**Comment by clearloop** [comment]
The compile error is caused by this:

https://github.com/darwinia-network/darwinia/blob/307b76c95ea3d2c3634f1e3716309cce1e0f7dd4/frame/chainrelay/eth/relay/src/tests.rs#L178

[darwinia/develop][0] added new test into `chainrelay/eth/relay`.

[0]: https://github.com/darwinia-network/darwinia/blob/develop/frame/chainrelay/eth/relay/src/tests.rs

**Comment by HackFisher** [comment]
> The compile error is caused by this:
> 
> https://github.com/darwinia-network/darwinia/blob/307b76c95ea3d2c3634f1e3716309cce1e0f7dd4/frame/chainrelay/eth/relay/src/tests.rs#L178
> 
> [darwinia/develop](https://github.com/darwinia-network/darwinia/blob/develop/frame/chainrelay/eth/relay/src/tests.rs) added new test into `chainrelay/eth/relay`.
> 
> I'll cherry pick [ffb88b8](https://github.com/darwinia-network/darwinia/commit/ffb88b83a09b71c846396526c589bbb97943f0bb) and update it together.

Understood, this is a diff introduced after your forking and have no conflicts with your PR.

So, "merge from develop" should always be suggested for pull requests.

**Comment by clearloop** [comment]
The  ffb88b83a09b71c846396526c589bbb97943f0bb has too many conflicts with this pr, I'll rebase and force push instead.

**Pull request merged**